crates/bindings-csharp/BSATN.Runtime/Builtins.cs

```diff
@@ -8,11 +8,152 @@ namespace SpacetimeDB;
+    /// (The endianness of the host system can be checked via System.BitConverter.IsLittleEndian.)
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="source"></param>
+    /// <param name="reverse"></param>
+    /// <returns></returns>
+    static T Read<T>(ReadOnlySpan<byte> source, bool reverse)
+        where T : struct
+    {
+        Debug.Assert(
+            source.Length == Marshal.SizeOf<T>(),
+            $"Error while reading ${typeof(T).FullName}: expected source span to be {Marshal.SizeOf<T>()} bytes long, but was {source.Length} bytes."
+        );
+
+        var result = MemoryMarshal.Read<T>(source);
+
+        if (reverse)
+        {
+            var resultSpan = MemoryMarshal.CreateSpan(ref result, 1);
+            MemoryMarshal.AsBytes(resultSpan).Reverse();
+        }
+
+        return result;
+    }
+
+    /// <summary>
+    /// Convert the passed T to a little-endian byte array.
+    /// The behavior of this method is independent of the endianness of the host, unlike MemoryMarshal.Read.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="source"></param>
+    /// <returns></returns>
+    public static byte[] AsBytesLittleEndian<T>(T source)
+        where T : struct => AsBytes(source, !BitConverter.IsLittleEndian);
+
+    /// <summary>
+    /// Convert the passed T to a big-endian byte array.
+    /// The behavior of this method is independent of the endianness of the host, unlike MemoryMarshal.Read.
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="source"></param>
+    /// <returns></returns>
+    public static byte[] AsBytesBigEndian<T>(T source)
+        where T : struct => AsBytes<T>(source, BitConverter.IsLittleEndian);
+
+    /// <summary>
+    /// Convert the passed T to a byte array, and optionally reverse the array before returning it.
+    /// If the output is not reversed, it will have the native endianness of the host system.
+    /// (The endianness of the host system can be checked via System.BitConverter.IsLittleEndian.)
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    /// <param name="source"></param>
+    /// <param name="reverse"></param>
+    /// <returns></returns>
+    static byte[] AsBytes<T>(T source, bool reverse)
+        where T : struct
+    {
+        var result = MemoryMarshal.AsBytes([source]).ToArray();
+        if (reverse)
+        {
+            Array.Reverse(result, 0, result.Length);
+        }
+        return result;
+    }
+
+    /// <summary>
+    /// Convert a hex string to a byte array.
+    /// </summary>
+    /// <param name="hex"></param>
+    /// <returns></returns>
+    public static byte[] StringToByteArray(string hex)
+    {
+        Debug.Assert(
+            hex.Length % 2 == 0,
+            $"Expected input string (\"{hex}\") to be of even length"
+        );
+
+        foreach (var c in hex)
+        {
+            Debug.Assert(
```

Comment:
I don't think this is necessary, `Convert.ToByte` should do it by itself. When I said "do we need a test", I really only meant a unit test :)